sdk/src/main/scala/io/horizen/account/sc2sc/AbstractCrossChainMessageProcessor.scala

```diff
@@ -0,0 +1,173 @@
+package io.horizen.account.sc2sc
+
+import com.google.common.primitives.{Bytes, Ints}
+import io.horizen.account.abi.ABIUtil.{METHOD_ID_LENGTH, getABIMethodId, getArgumentsFromData}
+import io.horizen.account.abi.{ABIDecoder, ABIEncodable, ABIListEncoder}
+import io.horizen.account.state.NativeSmartContractMsgProcessor.NULL_HEX_STRING_32
+import io.horizen.account.state.events.AddCrossChainMessage
+import io.horizen.account.state.{AccountStateView, BaseAccountStateView, ExecutionRevertedException, Message, NativeSmartContractMsgProcessor}
+import io.horizen.account.utils.ZenWeiConverter
+import io.horizen.cryptolibprovider.CryptoLibProvider
+import io.horizen.evm.Address
+import io.horizen.params.NetworkParams
+import io.horizen.sc2sc.{CrossChainMessage, CrossChainMessageHash, CrossChainMessageImpl, CrossChainProtocolVersion}
+import io.horizen.utils.ZenCoinsUtils
+import org.web3j.abi.TypeReference
+import org.web3j.abi.datatypes.{StaticStruct, Type}
+import org.web3j.abi.datatypes.generated.Uint32
+import sparkz.crypto.hash.Keccak256
+
+import java.math.BigInteger
+import java.util
+import scala.collection.JavaConverters.seqAsJavaListConverter
+
+trait CrossChainMessageProvider {
+  private[horizen] def getCrossChainMessages(epochNum: Int, view: BaseAccountStateView): Seq[CrossChainMessage]
+  private[horizen] def getCrossChainMessageHashEpoch(msgHash: CrossChainMessageHash, view: BaseAccountStateView): Option[Int]
+}
+abstract class AbstractCrossChainMessageProcessor(networkParams: NetworkParams) extends NativeSmartContractMsgProcessor with CrossChainMessageProvider {
```

Comment:
Consider that this class will be used/extended by application developers => must be Java class.
Same for `AbstractCrossChainRedeemMessageProcessor`